File: tokenrazor/core/scanner.py
"""CoT 冗余模式检测器。"""

import re
from typing import List


class RedundancyMatch:
    """单次冗余匹配。"""
    __slots__ = ("start", "end", "reason", "confidence")

    def __init__(self, start: int, end: int, reason: str, confidence: float):
        self.start = start
        self.end = end
        self.reason = reason
        self.confidence = confidence

    def __repr__(self):
        return f"Redundancy({self.start}:{self.end}, {self.reason}, conf={self.confidence:.2f})"


class Scanner:
    """扫描 CoT 文本中的冗余模式。"""

    # 填充短语模式（中英文）
    # 原则：宁可少杀，不可误杀。只移除确凿的过渡废话。
    FILLER_PHRASES = re.compile(
        r"(?i)(?:"
        r"let me (?:think|consider|check|verify|re-?evaluate) (?:about |on |this |that )?(?:step by step|carefully)?"
        r"|i (?:need to|should|must|will|could|would like to) (?:think about|consider|check|verify|analyze|look at|start by)"
        r"|let's (?:think|consider|work through|break down|go through)"
        r"|ok(?:ay)?,? (?:let|so|here)"
        r"|hmm,? (?:let|that|this)"
        r"|now,? let"
        r"|so,? let"
        r"|actually,? let"
        r"|wait,? let"
        r"|alright,? let"
        r"|as i (?:mentioned|said|noted)"
        # 中文填充
        r"|让我(?:再?[想想看]+|[来再](?:想想|看看|梳理一下|整理一下))"
        r"|首先,?我(?:们)?(?:先)?"
        r"|好了,?我(?:们)?"
        r"|好的,?我(?:们)?"
        r"|嗯,?我"
        r"|等等,?让我"
        r")"
    )

    # 平行枚举模式（中英文）
    # 检测 "方法一...方法二...方法三..." 这类枚举结构
    PARALLEL_ENUM_LINE = re.compile(
        r"(?i)^[\s]*"
        r"(?:"
        r"(?:方法|方案|方式|思路|途径|option|approach|method|way)"
        r"[\s]?"
        r"[一二三四五六七八九十1234567890、,]+[:：]"
        r"|[一二三四五六七八九十]+\.?[、.]?[\s]?[:：]?"
        r"|(?:option|approach|method|way)\s*[\s1234567890,:：]+"
        r")"
    )

    # 死胡同模式（中英文）
    DEAD_END_PATTERNS = re.compile(
        r"(?i)(?:"
        r"(?:that|this|the (?:above|previous|first)) (?:approach|method|way|path|idea|thought|option|direction) "
        r"(?:doesn't|didn't|won't|wouldn't) (?:work|make sense|fit|apply|help)"
        r"|let me (?:re-?approach|re-?try|re-?start|try again|try a different|take a different)"
        r"|scratch that"
        r"|disregard (?:that|the above|the previous)"
        # 中文死胡同
        r"|(?:方法[一二三四五六七八九十]?|思路[一二三四五六七八九十]?|方案[一二三四五六七八九十]?|方式)"
        r"(?:"
        r"(?:不太|不|并不|可能|也)?(?:可行|适用|正确|合适|好|对|行|妥当)"
        r"|有问题|有局限|有不足|有缺陷"
        r")"
        r"|换个(?:思路|方法|方向|方案)"
        r"|仔细想想,?(?:这个|这种|以上)?(?:方式?|方法|思路|方案)(?:不太|不|可能)?(?:对|妥|好|行|合适)"
        r"|算[了啦],?(?:这个|这条|这种方法|这个方案)?(?:放弃|不行|算了)"
        r")"
    )

    def scan_all(self, text: str, strategies: List[str] = None) -> List[RedundancyMatch]:
        """对文本执行全部扫描模式。

        Args:
            text: 待扫描文本
            strategies: 启用的策略列表，None 表示全部启用
        """
        matches: List[RedundancyMatch] = []

        if strategies is None or "filler" in strategies:
            matches.extend(self._scan_fillers(text))
        if strategies is None or "dead_end" in strategies:
            matches.extend(self._scan_dead_ends(text))
        if strategies is None or "parallel_enum" in strategies:
            matches.extend(self._scan_parallel_enums(text))

        # 排序去重
        matches.sort(key=lambda m: m.start)
        return self._deduplicate(matches)

    def _scan_fillers(self, text: str) -> List[RedundancyMatch]:
        matches = []
        for m in self.FILLER_PHRASES.finditer(text):
            matches.append(RedundancyMatch(m.start(), m.end(), "filler", 0.85))
        return matches

    def _scan_dead_ends(self, text: str) -> List[RedundancyMatch]:
        matches = []
        for m in self.DEAD_END_PATTERNS.finditer(text):
            sent_start = text.rfind("\n", 0, m.start())
            if sent_start == -1:
                sent_start = 0
            sent_end = text.find("\n", m.end())
            if sent_end == -1:
                sent_end = len(text)
            matches.append(
                RedundancyMatch(sent_start, sent_end, "dead_end", 0.8)
            )
        return matches

    def _scan_parallel_enums(self, text: str) -> List[RedundancyMatch]:
        """检测平行枚举分支并标记为可折叠。

        模式识别：连续的 "方法一/方案二/思路三/option 1/approach 2" 等枚举结构。
        策略：保留最后一个枚举块（通常是被选中的方案），标记前面的为冗余。
        """
        lines = text.split("\n")
        matches = []

        enum_block_start = -1
        enum_block_end = -1
        enum_count = 0
        char_offset = 0

        for i, line in enumerate(lines):
            stripped = line.strip()
            # 检测枚举行
            is_enum = bool(self._enum_marker(stripped))

            if is_enum:
                if enum_count == 0:
                    enum_block_start = char_offset
                enum_count += 1
                enum_block_end = char_offset + len(line) + 1  # +1 for newline
            else:
                if enum_count >= 3:
                    # 有一个完整的枚举块（至少3个枚举项）
                    # 标记除了最后一项以外的所有枚举项为冗余
                    # 但更精确的做法是按行标记
                    pass
                enum_count = 0

            char_offset += len(line) + 1

        # 第二次扫描：找连续枚举行（至少3行）
        text_copy = text
        enum_pattern = re.compile(
            r"(?:^[\s]*"
            r"(?:方法|方案|方式|思路|途径|option|approach|method|way)"
            r"[\s]?[一二三四五六七八九十1234567890、,]+[:：][^\n]*?\n?)"
            r"{3,}",
            re.IGNORECASE | re.MULTILINE,
        )

        for m in enum_pattern.finditer(text):
            block = m.group()
            block_lines = [l for l in block.split("\n") if l.strip()]
            if len(block_lines) >= 3:
                # 只保留最后一个枚举项，其余标记
                remove_end = m.start()
                for line_text in block_lines[:-1]:
                    line_end = remove_end + len(line_text) + 1
                    matches.append(RedundancyMatch(
                        remove_end, line_end, "parallel_enum", 0.75,
                    ))
                    remove_end = line_end

        return matches

    @staticmethod
    def _enum_marker(line: str) -> bool:
        """判断一行是否为枚举标记。"""
        return bool(
            re.search(
                r"(?i)^[\s]*"
                r"(?:"
                r"(?:方法|方案|方式|思路|途径|option|approach|method|way)"
                r"[\s]?[一二三四五六七八九十1234567890、,]+[:：]"
                r"(?:.*)"
                r")",
                line,
            )
            or re.search(r"^(?:第一|第二|第三|第四|首先|其次|再次|最后)[，,、]?", line)
            or re.search(r"^[一二三四五六七八九十]+\.?[、.]", line)
        )

    def _deduplicate(self, matches: List[RedundancyMatch]) -> List[RedundancyMatch]:
        """合并/去重重叠区间。"""
        if not matches:
            return []

        cleaned: List[RedundancyMatch] = [matches[0]]
        for m in matches[1:]:
            prev = cleaned[-1]
            if m.start >= prev.start and m.end <= prev.end:
                continue
            if m.start < prev.end:
                merged = RedundancyMatch(
                    prev.start, max(prev.end, m.end),
                    f"{prev.reason}+{m.reason}",
                    max(prev.confidence, m.confidence),
                )
                cleaned[-1] = merged
            else:
                cleaned.append(m)

        return cleaned

File: rust/src/scanner.rs
/// CoT 冗余模式扫描器
///
/// 中文差异化：中英文双语模式全覆盖

use regex::Regex;
use serde::Serialize;

/// 一次冗余匹配
#[derive(Debug, Clone, Serialize)]
pub struct RedundancyMatch {
    pub start: usize,
    pub end: usize,
    pub reason: String,
    pub confidence: f64,
}

impl RedundancyMatch {
    pub fn new(start: usize, end: usize, reason: &str, confidence: f64) -> Self {
        Self { start, end, reason: reason.to_string(), confidence }
    }
}

/// CoT 模式扫描器
pub struct Scanner {
    filler_re: Regex,
    dead_end_re: Regex,
    enum_line_re: Regex,
}

impl Default for Scanner {
    fn default() -> Self {
        Self::new()
    }
}

impl Scanner {
    pub fn new() -> Self {
        // 填充短语（中英文）
        let filler_re = Regex::new(
            r"(?i)(?:let me (?:think|consider|check|verify|re-?evaluate) (?:about |on |this |that )?(?:step by step|carefully)?|i (?:need to|should|must|will|could|would like to) (?:think about|consider|check|verify|analyze|look at|start by)|let's (?:think|consider|work through|break down|go through)|ok(?:ay)?,? (?:let|so|here)|hmm,? (?:let|that|this)|now,? let|so,? let|actually,? let|wait,? let|alright,? let|as i (?:mentioned|said|noted)|让我(?:再?[想想看]+|[来再](?:想想|看看|梳理一下|整理一下))|首先,?我(?:们)?(?:先)?|好了,?我(?:们)?|好的,?我(?:们)?|嗯,?我|等等,?让我)"
        ).expect("filler regex");

        // 死胡同模式（中英文）
        let dead_end_re = Regex::new(
            r"(?i)(?:(?:that|this|the (?:above|previous|first)) (?:approach|method|way|path|idea|thought|option|direction) (?:doesn't|didn't|won't|wouldn't) (?:work|make sense|fit|apply|help)|let me (?:re-?approach|re-?try|re-?start|try again|try a different|take a different)|scratch that|disregard (?:that|the above|the previous)|(?:方法|思路|方案|方式)(?:(?:不太|不|并不|可能|也)?(?:可行|适用|正确|合适|好|对|行|妥当)|有问题|有局限|有不足|有缺陷)|换个(?:思路|方法|方向|方案)|仔细想想,?(?:这个|这种|以上)?(?:方式?|方法|思路|方案)(?:不太|不|可能)?(?:对|妥|好|行|合适)|算[了啦],?(?:这个|这条|这种方法|这个方案)?(?:放弃|不行|算了)|这个方法不行|此路不通)"
        ).expect("dead_end regex");

        // 枚举行检测
        let enum_line_re = Regex::new(
            r"(?i)(?:(?:方法|方案|方式|思路|途径|option|approach|method|way)\s?[一二三四五六七八九十\d、,]+[:：])"
        ).expect("enum_line regex");

        Self { filler_re, dead_end_re, enum_line_re }
    }

    /// 扫描填充短语
    pub fn scan_fillers(&self, text: &str) -> Vec<RedundancyMatch> {
        self.filler_re.find_iter(text)
            .map(|m| RedundancyMatch::new(m.start(), m.end(), "filler", 0.85))
            .collect()
    }

    /// 扫描死胡同
    pub fn scan_dead_ends(&self, text: &str) -> Vec<RedundancyMatch> {
        self.dead_end_re.find_iter(text)
            .map(|m| {
                let sent_start = text[..m.start()].rfind('\n').map(|p| p + 1).unwrap_or(0);
                let sent_end = text[m.end()..].find('\n')
                    .map(|p| m.end() + p)
                    .unwrap_or(text.len());
                RedundancyMatch::new(sent_start, sent_end, "dead_end", 0.80)
            })
            .collect()
    }

    /// 扫描平行枚举
    pub fn scan_parallel_enums(&self, text: &str) -> Vec<RedundancyMatch> {
        let lines: Vec<&str> = text.lines().collect();
        let mut enum_indices: Vec<(usize, usize)> = Vec::new();

        let mut char_offset = 0;
        for (i, line) in lines.iter().enumerate() {
            if self.enum_line_re.is_match(line.trim()) {
                enum_indices.push((i, char_offset));
            }
            char_offset += line.len() + 1;
        }

        if enum_indices.len() < 3 {
            return vec![];
        }

        let mut sequences: Vec<(usize, usize, usize, usize)> = Vec::new();
        let mut seq_start = 0;

        for j in 1..enum_indices.len() {
            let gap = enum_indices[j].0 - enum_indices[j - 1].0;
            if gap > 10 {
                if j - seq_start >= 3 {
                    let first = enum_indices[seq_start];
                    let last = enum_indices[j - 1];
                    sequences.push((seq_start, j - 1, first.1, last.1));
                }
                seq_start = j;
            }
        }
        if enum_indices.len() - seq_start >= 3 {
            let first = enum_indices[seq_start];
            let last = enum_indices[enum_indices.len() - 1];
            sequences.push((seq_start, enum_indices.len() - 1, first.1, last.1));
        }

        let mut matches = Vec::new();
        for (seq_s, seq_e, _, _) in sequences {
            for k in seq_s..seq_e {
                let block_start = enum_indices[k].1;
                let block_end = if k + 1 < enum_indices.len() {
                    enum_indices[k + 1].1
                } else {
                    text.len()
                };
                matches.push(RedundancyMatch::new(
                    block_start, block_end.saturating_sub(1), "parallel_enum", 0.75,
                ));
            }
        }
        matches
    }

    /// 执行全部扫描
    pub fn scan_all(&self, text: &str, strategies: &[&str]) -> Vec<RedundancyMatch> {
        let mut matches = Vec::new();

        if strategies.is_empty() || strategies.contains(&"filler") {
            matches.extend(self.scan_fillers(text));
        }
        if strategies.contains(&"dead_end") {
            matches.extend(self.scan_dead_ends(text));
        }
        if strategies.contains(&"parallel_enum") {
            matches.extend(self.scan_parallel_enums(text));
        }

        matches.sort_by(|a, b| a.start.cmp(&b.start));
        self.deduplicate(&matches)
    }

    fn deduplicate(&self, matches: &[RedundancyMatch]) -> Vec<RedundancyMatch> {
        if matches.is_empty() {
            return vec![];
        }

        let mut cleaned: Vec<RedundancyMatch> = vec![matches[0].clone()];
        for m in matches[1..].iter() {
            let prev = &cleaned[cleaned.len() - 1];
            if m.start >= prev.start && m.end <= prev.end {
                continue;
            }
            if m.start < prev.end {
                cleaned.push(RedundancyMatch::new(
                    prev.start, prev.end.max(m.end),
                    &format!("{}+{}", prev.reason, m.reason),
                    prev.confidence.max(m.confidence),
                ));
            } else {
                cleaned.push(m.clone());
            }
        }
        cleaned
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scan_fillers_english() {
        let s = Scanner::new();
        let m = s.scan_fillers("Let me think about this step by step.");
        assert!(!m.is_empty());
    }

    #[test]
    fn test_scan_fillers_chinese() {
        let s = Scanner::new();
        let m = s.scan_fillers("让我想想这个问题。");
        assert!(!m.is_empty(), "中文填充应被检测");
    }

    #[test]
    fn test_scan_dead_end_chinese() {
        let s = Scanner::new();
        let text = "这个方法不行。让我换一个思路。";
        let m = s.scan_dead_ends(text);
        assert!(!m.is_empty());
    }

    #[test]
    fn test_scan_parallel_enum() {
        let s = Scanner::new();
        let text = "方案一：A。\n方案二：B。\n方案三：C。\n我选三。";
        let m = s.scan_parallel_enums(text);
        assert!(!m.is_empty());
    }

    #[test]
    fn test_scan_all_returns_matches() {
        let s = Scanner::new();
        let text = "让我想想。方法一。\n方法二。\n方法三。\n方法不行。答案：C";
        let m = s.scan_all(text, &["filler", "dead_end", "parallel_enum"]);
        assert!(!m.is_empty());
    }

    #[test]
    fn test_no_false_positives() {
        let s = Scanner::new();
        let text = "这是正常对话。答案：42。";
        let m = s.scan_all(text, &["filler", "dead_end", "parallel_enum"]);
        assert!(m.len() < 3);
    }
}
